File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/domain/converter/UnitConverterTest.java
package com.tarbadev.witchcraft.domain.converter;

import com.tarbadev.witchcraft.domain.converter.UnitConverter;
import org.junit.Before;
import org.junit.Test;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.AbstractMap;
import java.util.Map;

import static org.assertj.core.api.Java6Assertions.assertThat;

public class UnitConverterTest {
  private UnitConverter subject;

  @Before
  public void setUp() {
    subject = new UnitConverter();
  }

  @Test
  public void convert_OzToLb() {
    Double oz = 16.0;
    Double expectedLb = 1.0;

    assertThat(subject.convert(oz, UnitConverter.UnitName.OZ.getName(), UnitConverter.UnitName.LB.getName())).isEqualTo(expectedLb);
  }

  @Test
  public void convert_tspToCup() {
    Double tsp = 48.0;
    Double cup = 1.0;

    assertThat(subject.convert(tsp, UnitConverter.UnitName.TSP.getName(), UnitConverter.UnitName.CUP.getName())).isEqualTo(cup);
  }

  @Test
  public void convertToHighestUnit_MlToOz() {
    Double ml = 1.0;
    Double oz = 0.033814;

    Map.Entry<String, Double> expectedResult = new AbstractMap.SimpleEntry<>(UnitConverter.UnitName.OZ.getName(), oz);
    Map.Entry<String, Double> returnedValue = subject.convertToHighestUnit(ml, UnitConverter.UnitName.ML.getName(), UnitConverter.UnitName.OZ.getName());
    returnedValue.setValue(new BigDecimal(returnedValue.getValue()).setScale(6, RoundingMode.HALF_UP).doubleValue());

    assertThat(returnedValue).isEqualTo(expectedResult);
  }

  @Test
  public void convertToHighestUnit_OzToMlReturnsOz() {
    Double oz = 0.033814;

    Map.Entry<String, Double> expectedResult = new AbstractMap.SimpleEntry<>(UnitConverter.UnitName.OZ.getName(), oz);
    Map.Entry<String, Double> returnedValue = subject.convertToHighestUnit(oz, UnitConverter.UnitName.OZ.getName(), UnitConverter.UnitName.ML.getName());
    returnedValue.setValue(new BigDecimal(returnedValue.getValue()).setScale(6, RoundingMode.HALF_UP).doubleValue());

    assertThat(returnedValue).isEqualTo(expectedResult);
  }

  @Test
  public void convertToHighestUnit_TspToCup() {
    Double tsp = 3.0;
    Double cup = 0.0625;

    Map.Entry<String, Double> expectedResult = new AbstractMap.SimpleEntry<>(UnitConverter.UnitName.CUP.getName(), cup);
    Map.Entry<String, Double> returnedValue = subject.convertToHighestUnit(tsp, UnitConverter.UnitName.TSP.getName(), UnitConverter.UnitName.CUP.getName());

    assertThat(returnedValue).isEqualTo(expectedResult);
  }

  @Test
  public void convertToHighestUnit_TbspToCup() {
    Double tbsp = 10.0;
    Double cup = 0.625;

    Map.Entry<String, Double> expectedResult = new AbstractMap.SimpleEntry<>(UnitConverter.UnitName.CUP.getName(), cup);
    Map.Entry<String, Double> returnedValue = subject.convertToHighestUnit(tbsp, UnitConverter.UnitName.TBSP.getName(), UnitConverter.UnitName.CUP.getName());

    assertThat(returnedValue).isEqualTo(expectedResult);
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/converter/UnitConverterTest.kt
package com.tarbadev.witchcraft.converter

import com.tarbadev.witchcraft.converter.UnitConverter.UnitName.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import tech.units.indriya.ComparableQuantity
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.AbstractMap.SimpleEntry
import javax.measure.quantity.Volume

class UnitConverterTest {
  private val unitConverter: UnitConverter = UnitConverter()

  @Test
  fun convert_OzToLb() {
    val oz = 16.0
    val expectedLb = 1.0

    assertEquals(expectedLb, unitConverter.convert(oz, OZ.value, LB.value))
  }

  @Test
  fun convert_tspToCup() {
    val tsp = 48.0
    val cup = 1.0

    assertEquals(cup, unitConverter.convert(tsp, TSP.value, CUP.value))
  }

  @Test
  fun convert_tspToTbsp() {
    val tsp = 3.0
    val tbsp = 1.0

    assertEquals(tbsp, unitConverter.convert(tsp, TSP.value, TBSP.value))
  }

  @Test
  fun convertToHighestUnit_MlAndFlOz_ReturnsFlOz() {
    val input1 = 1.milliliter
    val input2 = 1.fluid_ounce

    val returnedValue = setScale(unitConverter.convertToHighestUnit(input1, input2), 6)
    val expectedResult = setScale(1.033814.fluid_ounce, 6)

    assertThat(returnedValue).isEqualTo(expectedResult)
  }

  @Test
  fun convertToHighestUnit_OzToMlReturnsOz() {
    val oz = 0.033814

    val expectedResult = SimpleEntry<String, Double>(OZ.value, oz)
    val returnedValue = unitConverter.convertToHighestUnit(oz, OZ.value, ML.value)
    returnedValue.setValue(BigDecimal(returnedValue.value).setScale(6, RoundingMode.HALF_UP).toDouble())

    assertEquals(expectedResult, returnedValue)
  }

  @Test
  fun convertToHighestUnit_TspToCup() {
    val tsp = 3.0
    val cup = 0.0625

    val expectedResult = SimpleEntry<String, Double>(CUP.value, cup)
    val returnedValue = unitConverter.convertToHighestUnit(tsp, TSP.value, CUP.value)

    assertEquals(expectedResult, returnedValue)
  }

  @Test
  fun convertToHighestUnit_TbspToCup() {
    val tbsp = 10.0
    val cup = 0.625

    val expectedResult = SimpleEntry<String, Double>(CUP.value, cup)
    val returnedValue = unitConverter.convertToHighestUnit(tbsp, TBSP.value, CUP.value)

    assertEquals(expectedResult, returnedValue)
  }

  @Test
  fun convertToHighestUnit_TspToTbsp() {
    val tsp = 3.0
    val tbsp = 1.0

    val expectedResult = SimpleEntry<String, Double>(TBSP.value, tbsp)
    val returnedValue = unitConverter.convertToHighestUnit(tsp, TSP.value, TBSP.value)

    assertEquals(expectedResult, returnedValue)
  }

  private fun setScale(number: ComparableQuantity<Volume>, scale: Int): BigDecimal {
    return BigDecimal(number.value.toDouble()).setScale(scale, RoundingMode.HALF_DOWN)
  }
}